src/v/kafka/server/write_at_offset_stm.cc

```diff
@@ -0,0 +1,270 @@
+    case errc::replicate_exception:
+        return "Exception thrown during replication";
+    case errc::invalid_batch_type:
+        return "Invalid batch type. Offset translated batches are not "
+               "supported by write at offset state machine";
+    }
+    __builtin_unreachable();
+}
+
+const std::error_category& write_at_offset_stm::error_category() noexcept {
+    static errc_category e;
+    return e;
+}
+
+std::error_code write_at_offset_stm::make_error_code(errc e) noexcept {
+    return {static_cast<int>(e), error_category()};
+}
+
+write_at_offset_stm::write_at_offset_stm(
+  raft::consensus* raft,
+  ss::logger& logger,
+  storage::kvstore& kvstore,
+  std::vector<model::record_batch_type> offset_translated_batches)
+  : raft::persisted_stm<raft::kvstore_backed_stm_snapshot>(
+      cluster::write_at_offset_stm_snapshot, logger, raft, kvstore)
+  , _offset_translated_batches(std::move(offset_translated_batches))
+  , _sync_lock("w-at-offset") {}
+
+raft::replicate_stages write_at_offset_stm::replicate(
+  model::record_batch batch,
+  kafka::offset expected_base_offset,
+  std::optional<kafka::offset> prev_log_offset,
+  model::timeout_clock::duration timeout) {
+    ss::promise<> enqueued_promise;
+    auto f = enqueued_promise.get_future();
+    return raft::replicate_stages{
+      std::move(f),
+      do_replicate(
+        std::move(batch),
+        expected_base_offset,
+        prev_log_offset,
+        timeout,
+        std::move(enqueued_promise))};
+}
+
+ss::future<result<raft::replicate_result>> write_at_offset_stm::do_replicate(
+  model::record_batch batch,
+  kafka::offset expected_base_offset,
+  std::optional<kafka::offset> last_log_offset,
+  model::timeout_clock::duration timeout,
+  ss::promise<> enqueued_promise) {
+    // offset translated batches are not supported by write at offset state
+    // machine
+    if (is_offset_translated_batch(batch)) [[unlikely]] {
+        enqueued_promise.set_value();
+        co_return make_error_code(errc::invalid_batch_type);
+    }
+    /**
+     * We use lock to serialize the sync calls, as sync is a scheduling point
+     * The units are released after the request is enqueued in raft.
+     */
+    auto u = co_await _sync_lock.get_units();
+    auto sync_result = co_await sync(timeout);
+    if (!sync_result) {
+        _inflight_last_offset.reset();
+        enqueued_promise.set_value();
+        co_return raft::errc::not_leader;
+    }
+
+    auto stm_last_offset = expected_last_offset();
+    vlog(
+      _log.trace,
+      "Requested replicate at offset: {} with previous log offset: {}. "
+      "stm last offset: {}",
+      expected_base_offset,
+      last_log_offset,
+      stm_last_offset);
+    const auto stm_expected_next_offset = kafka::next_offset(stm_last_offset);
+    chunked_vector<model::record_batch> batches;
+    /**
+     * When prev log offset is set, we are dealing with a gap
+     */
+    if (last_log_offset) {
+        batches.reserve(2);
```

Comment:
This is a bit confusing here, maybe it's better to have right before the `move`? Also it threw me off that `make_ghost_batches` could make multiple and we always reserve 2 only, but that because `make_ghost_batches` only returns more if the delta is >= max int32, which is highly, highly unlikely.